component/azstorage/datalake.go

```diff
@@ -553,7 +553,40 @@ func (dl *Datalake) ReadInBuffer(name string, offset int64, len int64, data []by
 
 // WriteFromFile : Upload local file to file
 func (dl *Datalake) WriteFromFile(name string, metadata map[string]*string, fi *os.File) (err error) {
-	return dl.BlockBlob.WriteFromFile(name, metadata, fi)
+	// File in DataLake may have permissions and ACL set. Just uploading the file will override them.
+	// So, we need to get the existing permissions and ACL and set them back after uploading the file.
+
+	var acl string = ""
+	var fileClient *file.Client = nil
+
+	if dl.Config.preserveACL {
+		fileClient = dl.Filesystem.NewFileClient(filepath.Join(dl.Config.prefixPath, name))
+		resp, err := fileClient.GetAccessControl(context.Background(), nil)
```

Comment:
File ACL should be present locally before upload, I am guessing.
Do we require to fetch it from storage(backend datalake)?

Reply:
Permissions are present in local like the "rwx" triplet but on blob you can have ACLs as well like giving permissions to a specific MSI/SPN/Identity. These are not replicated on local system and we do not cache them anywhere in blobfuse as well. This is the reason we need to get the latest ACLs and permissions from storage before we upload the file. Uploading a file using PutBlob is just resetting all the applied permissions and ACLs.
That being said, this has an extra cost of 2 additional REST calls. By default we were not doing this and customers who are ok with that may not want to take this additional cost. Hence we have given a flag to configure.